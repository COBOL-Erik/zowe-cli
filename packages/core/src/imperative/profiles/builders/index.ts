/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

export * from "./CompleteProfilesGroupBuilder";
export * from "./ProfilesCommandBuilder";
export * from "./ProfilesCreateCommandBuilder";
export * from "./ProfilesDeleteCommandBuilder";
export * from "./ProfilesListCommandBuilder";
export * from "./ProfilesSetCommandBuilder";
export * from "./ProfilesShowDependenciesCommandBuilder";
export * from "./ProfilesUpdateCommandBuilder";
export * from "./ProfilesValidateCommandBuilder";
