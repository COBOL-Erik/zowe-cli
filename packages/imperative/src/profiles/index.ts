/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

export * from "./src/constants/ProfilesConstants";

export * from "./src/doc";

export * from "./src/doc/definition/IMetaProfile";
export * from "./src/doc/definition/IProfile";
export * from "./src/doc/definition/IProfileDependency";
export * from "./src/doc/definition/IProfileProperty";
export * from "./src/doc/definition/IProfileSchema";

export * from "./src/doc/parms/IProfileManager";
export * from "./src/doc/response/IProfileLoaded";

export * from "./src/utils/ProfileIO";
export * from "./src/utils/ProfileUtils";
export * from "./src/utils";

export * from "./src/validation/api/ProfileValidator";

export * from "./src/validation/doc/IProfileValidationPlan";
export * from "./src/validation/doc/IProfileValidationReport";
export * from "./src/validation/doc/IProfileValidationTask";
export * from "./src/validation/doc/IProfileValidationTaskResult";

