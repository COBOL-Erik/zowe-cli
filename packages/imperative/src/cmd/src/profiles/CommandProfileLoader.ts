/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { Arguments } from "yargs";
import { ICommandDefinition } from "../doc/ICommandDefinition";
import { IProfile, IProfileLoaded } from "../../../profiles";
import { CommandProfiles } from "./CommandProfiles";
import { inspect } from "util";
import { ICommandProfileLoaderParms } from "../doc/profiles/parms/ICommandProfileLoaderParms";
import { Logger } from "../../../logger";
import { ImperativeExpect } from "../../../expect";

/**
 * The command processor profile loader loads all profiles that are required (or optional) given a command
 * definitions requirements. It returns the CommandProfiles object (which contains the map and getters for the
 * command handlers usage).
 * @internal
 * @class CommandProfileLoader
 * @deprecated Use the `V1ProfileRead` class if you still need to read V1 profiles
 */
export class CommandProfileLoader {
    /**
     * Create a new instance of the profile loader
     * @static
     * @param {ICommandProfileLoaderParms} parms - contains command definition and logger
     * @returns
     * @memberof CommandProfileLoader
     */
    public static loader(parms: ICommandProfileLoaderParms) {
        // eslint-disable-next-line deprecation/deprecation
        return new CommandProfileLoader(parms.commandDefinition, parms.logger || Logger.getImperativeLogger());
    }

    /**
     * The input command definition for the command being issued.
     * @private
     * @type {ICommandDefinition}
     * @memberof CommandProfileLoader
     */
    private mCommandDefinition: ICommandDefinition;

    /**
     * Logger - supplied on the constructor - but defaults to the Imperative logger.
     * @private
     * @type {Logger}
     * @memberof CommandProfileLoader
     */
    private mLog: Logger;

    /**
     * Creates an instance of CommandProfileLoader.
     * @param {ICommandDefinition} commandDefinition - The input command definition for the command being issued.
     * @param {any} [logger=Logger.getImperativeLogger()] - A log4js instance
     * @memberof CommandProfileLoader
     */
    constructor(commandDefinition: ICommandDefinition, logger = Logger.getImperativeLogger()) {
        const err: string = "Could not construct the profile loader.";
        ImperativeExpect.toNotBeNullOrUndefined(commandDefinition, `${err} No command definition supplied.`);
        this.mCommandDefinition = commandDefinition;
        ImperativeExpect.toBeEqual(logger instanceof Logger, true, `${err} The "logger" supplied is not of type Logger.`);
        this.mLog = logger;
        this.log.trace(`Profile loader created for command: ${commandDefinition.name}`);
    }

    /**
     * Load the profiles for the command - the command arguments are supplied to grab the profile names from
     * the arguments supplied by the user.
     * @param {Arguments} commandArguments - The command arguments supplied on this command invocation (Yargs style)
     * @returns {Promise<CommandProfiles>} - The promise is fulfilled with the map object OR rejected with an
     * Imperative error
     * @memberof CommandProfileLoader
     */
    // eslint-disable-next-line deprecation/deprecation
    public async loadProfiles(commandArguments: Arguments): Promise<CommandProfiles> {
        // Validate parms
        ImperativeExpect.toNotBeNullOrUndefined(commandArguments, `Could not load profiles. No command arguments supplied.`);

        // Log the API call
        this.log.info(`Request to load profiles for command: ${this.definition.name}...`);
        this.log.trace(`Profile load arguments supplied:\n${inspect(commandArguments, {depth: null})}`);

        // Create the map that eventually will be returned
        const profileMap: Map<string, IProfile[]> = new Map<string, IProfile[]>();
        const profileMetaMap: Map<string, IProfileLoaded[]> = new Map<string, IProfileLoaded[]>();

        // We no longer read V1 profile files, so just return empty maps
        // eslint-disable-next-line deprecation/deprecation
        return new CommandProfiles(profileMap, profileMetaMap);
    }

    /**
     * Accessor for the command definition document
     * @readonly
     * @private
     * @type {ICommandDefinition}
     * @memberof CommandProfileLoader
     */
    private get definition(): ICommandDefinition {
        return this.mCommandDefinition;
    }

    /**
     * Accessor for the logging object
     * @readonly
     * @private
     * @type {Logger}
     * @memberof CommandProfileLoader
     */
    private get log(): Logger {
        return this.mLog;
    }
}
