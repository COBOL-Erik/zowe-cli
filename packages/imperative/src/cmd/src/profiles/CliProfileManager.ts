/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { BasicProfileManager } from "../../../profiles";
import { ImperativeError } from "../../../error";
import { ICommandProfileTypeConfiguration } from "../doc/profiles/definition/ICommandProfileTypeConfiguration";
import { IProfileLoaded } from "../../../profiles/src/doc";
import { ICliLoadAllProfiles } from "../doc/profiles/parms/ICliLoadAllProfiles";

/**
 * A profile management API compatible with transforming command line arguments into
 * profiles
 * @internal
 */
export class CliProfileManager extends BasicProfileManager<ICommandProfileTypeConfiguration> {

    /**
     * NOTE: This is just a copy of BasicProfileManager.loadAll
     * REASON: We needed the Abstract profile manager to call the CLI profile manager to handle loading of secure properties
     * Loads all profiles from every type. Profile types are determined by reading all directories within the
     * profile root directory.
     * @returns {Promise<IProfileLoaded[]>} - The list of all profiles for every type
     */
    public async loadAll(params?: ICliLoadAllProfiles): Promise<IProfileLoaded[]> {
        this.log.trace(`Loading all profiles for type "${this.profileType}"...`);
        // Load all the other profiles for other types
        const loadAllProfiles: any[] = [];

        // Load only the profiles for the type if requested
        if (params != null && params.typeOnly) {
            const names: string[] = this.getAllProfileNames();
            for (const name of names) {
                loadAllProfiles.push(this.load({
                    name,
                    failNotFound: true,
                    loadDependencies: false,
                    noSecure: params.noSecure
                }));
            }
        } else {

            // Otherwise, load all profiles of all types
            for (const typeConfig of this.profileTypeConfigurations) {
                const typeProfileManager = new CliProfileManager({
                    profileRootDirectory: this.profileRootDirectory,
                    typeConfigurations: this.profileTypeConfigurations,
                    type: typeConfig.type,
                    logger: this.log,
                    loadCounter: this.loadCounter
                });

                // Get all the profile names for the type and attempt to load every one
                const names: string[] = typeProfileManager.getAllProfileNames();
                for (const name of names) {
                    this.log.debug(`Loading profile "${name}" of type "${typeConfig.type}".`);
                    loadAllProfiles.push(typeProfileManager.load({
                        name,
                        failNotFound: true,
                        loadDependencies: false,
                        noSecure: (params != null) ? params.noSecure : undefined
                    }));
                }
            }
        }

        // Construct the full list for return
        let allProfiles: IProfileLoaded[] = [];
        try {
            this.log.trace(`Awaiting all loads...`);
            const theirProfiles = await Promise.all(loadAllProfiles);
            for (const theirs of theirProfiles) {
                allProfiles = allProfiles.concat(theirs);
            }
            this.log.trace(`All loads complete.`);
        } catch (e) {
            this.log.error(e.message);
            throw new ImperativeError({msg: e.message, additionalDetails: e.additionalDetails, causeErrors: e});
        }

        return allProfiles;
    }
}
