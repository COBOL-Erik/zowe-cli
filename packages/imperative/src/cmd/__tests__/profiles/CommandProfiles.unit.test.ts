/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

/* eslint-disable deprecation/deprecation */

import { IProfile, IProfileLoaded } from "../../../profiles";
import { CommandProfiles } from "../../src/profiles/CommandProfiles";
import { ImperativeError } from "../../../error";

const BANANA_PROFILE_TYPE: string = "banana";
const STRAWBERRY_PROFILE_TYPE: string = "strawberry";

describe("Command Profiles", () => {
    it("should should allow us to create an instance", () => {
        let caughtError;
        try {
            const profiles = new CommandProfiles(new Map<string, IProfile[]>());
        } catch (error) {
            caughtError = error;
        }
        expect(caughtError).toBeUndefined();
    });

    it("should allow us to create an instance with map values", () => {
        const map = new Map<string, IProfile[]>();
        map.set(STRAWBERRY_PROFILE_TYPE, [{
            name: "great",
            type: STRAWBERRY_PROFILE_TYPE,
            age: 1
        }, {
            name: "awesome",
            type: STRAWBERRY_PROFILE_TYPE,
            age: 2
        }]);
        const metaMap = new Map<string, IProfileLoaded[]>();
        metaMap.set(STRAWBERRY_PROFILE_TYPE, [{
            name: "great",
            type: STRAWBERRY_PROFILE_TYPE,
            profile: {
                age: 1
            },
            message: "just right",
            failNotFound: false
        },
        {
            name: "gross",
            type: STRAWBERRY_PROFILE_TYPE,
            profile: {
                age: 3
            },
            message: "too old",
            failNotFound: false
        }]);

        const profiles = new CommandProfiles(map, metaMap);
        expect(profiles).toMatchSnapshot();
    });

    it("should detect missing parameters", () => {
        let error;
        try {
            new CommandProfiles(undefined);
        } catch (e) {
            error = e;
        }
        expect(error).toBeDefined();
        expect(error instanceof ImperativeError).toBe(true);
        expect(error.message).toMatchSnapshot();
    });

    it("should detect that the parameters are not a map", () => {
        let error;
        try {
            const map = { not: "a-map" };
            new CommandProfiles(map as any);
        } catch (e) {
            error = e;
        }
        expect(error).toBeDefined();
        expect(error instanceof ImperativeError).toBe(true);
        expect(error.message).toMatchSnapshot();
    });
});
