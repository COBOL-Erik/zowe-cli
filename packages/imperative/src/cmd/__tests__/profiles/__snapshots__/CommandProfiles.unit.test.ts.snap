// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Command Profiles should allow us to create an instance with map values 1`] = `
CommandProfiles {
  "mMap": Map {
    "strawberry" => Array [
      Object {
        "age": 1,
        "name": "great",
        "type": "strawberry",
      },
      Object {
        "age": 2,
        "name": "awesome",
        "type": "strawberry",
      },
    ],
  },
  "mMetaMap": Map {
    "strawberry" => Array [
      Object {
        "failNotFound": false,
        "message": "just right",
        "name": "great",
        "profile": Object {
          "age": 1,
        },
        "type": "strawberry",
      },
      Object {
        "failNotFound": false,
        "message": "too old",
        "name": "gross",
        "profile": Object {
          "age": 3,
        },
        "type": "strawberry",
      },
    ],
  },
}
`;

exports[`Command Profiles should detect missing parameters 1`] = `"Expect Error: Command Profiles Internal Error: No map was supplied."`;

exports[`Command Profiles should detect that the parameters are not a map 1`] = `"Expect Error: Command Profiles Internal Error: The \\"map\\" supplied is not an instance of a map."`;
